SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLoadBalancingHandler.h

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineClasses/SpatialActorChannel.h"
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "Interop/SpatialSender.h"
+#include "LoadBalancing/AbstractLBStrategy.h"
+#include "LoadBalancing/OwnershipLockingPolicy.h"
+#include "Schema/SpatialDebugging.h"
+#include "Utils/SpatialActorUtils.h"
+
+// Template class to handle LoadBalancing for a collection of actors.
+// Since different systems (NetDriver, ReplicationGraph) have different types for actor collection
+// and different ways to declare dependencies between actors, a CRTP pattern is used to have the
+// implementation class provide access to these information.
+// The methods to provide are :
+//  - GetActorsBeingReplicated() -> returns a range-for compatible iterator over AActor being replicated this frame
+//  - AddActorToReplicate(AActor* Actor) -> inform that an additional actor should be replicated this frame (a dependent actor)
+//  - RemoveAdditionalActor(AActor* Actor) -> inform that an actor we signaled as needing to replicate was encountered in the list of replicated actor, and does not need additional handling
+//  - GetDependentActors(AActor* Actor) -> returns a range-for compatible iterator over AActor that depends on the given actor (and should be migrated together)
+template <typename Implementation>
+class TSpatialLoadBalancingHandler
+{
+public:
+
+	TSpatialLoadBalancingHandler(USpatialNetDriver* InNetDriver)
+		: NetDriver(InNetDriver)
+	{
+
+	}
+
+	// Iterates over the list of actors to replicate, to check if they should migrate to another worker
+	// and collects additional actors to replicate if needed.
+	void HandleLoadBalancing()
+	{
+		TSet<FNetworkObjectInfo*> AdditionalObjectsToConsider;
+
+		for (AActor* Actor : static_cast<Implementation*>(this)->GetActorsBeingReplicated())
+		{
+			Worker_EntityId EntityId = NetDriver->PackageMap->GetEntityIdFromObject(Actor);
+			if (EntityId == SpatialConstants::INVALID_ENTITY_ID)
+			{
+				continue;
+			}
+
+			if (!Actor->HasAuthority())
+			{
+				continue;
+			}
```

Comment:
Can we check both of these - should they ever happen?

Reply:
First one can happen, if it's the first replication frame of an actor, but that might not be a reason to bail out actually.
Second one, I have seen actors being pulled in the replication loop without auth on occasion (and triggering https://improbableio.atlassian.net/browse/UNR-3757)

Reply:
In BuildConsiderList we skip non-auth Actors so I think we should check, and investigate if it fails

Reply:
Actually, it could be a case of how the previous recursive migration worked 🤔
Load balancing a group of actor once the first one in the hierarchy was encountered, and one of these actors being encountered later in the list.
Although it happens very infrequently while it should had a 50% chance of happening.

Reply:
> Load balancing a group of actor once the first one in the hierarchy was encountered, and one of these actors being encountered later in the list.

Double processing the same Actor sounds like something we'd want to avoid anyway - can we fix doing that?

>First one can happen, if it's the first replication frame of an actor

If it's the first replication frame, the previous call to ProcessPrioritizedActor should assign an entity ID when it creates an channel - is it possible this doesn't happen?

Reply:
> Double processing the same Actor sounds like something we'd want to avoid anyway - can we fix doing that?
Yes, the current PR would prevent that.

> If it's the first replication frame, the previous call to ProcessPrioritizedActor should assign an entity ID when it creates an channel - is it possible this doesn't happen?
HandleLoadBalancing is called before ProcessPrioritizedActor (in order to collect the additional actors to process). So that would be a change compared to how it previously happened.
We could do it in two steps ? Replicate, Load balance, and replicate additional actors, but that seems complicated.
So that's a change that I did nor foresee. Load balancing cannot happen on the first replication frame anymore. But from your earlier change that the spawner worker should have authority, that might not be an issue.